src/V3MergeCond.cpp

```diff
@@ -51,71 +51,347 @@
-    // - It contains an impure expression
-    // - It contains an LValue referencing the condition
-    Mergeable operator()(const AstNode* node) {
-        m_condAssign = false;
-        m_impure = false;
-        iterateChildrenConst(const_cast<AstNode*>(node));
-        if (m_impure) {  // Impure is stronger than cond assign
-            return Mergeable::NO_IMPURE;
-        } else if (m_condAssign) {
-            return Mergeable::NO_COND_ASSIGN;
+    void analyseNode(AstNode* nodep) {
+        if (m_propsp && !nodep->isPure()) m_propsp->m_isFence = true;
+        iterateChildrenConst(nodep);
+    }
+
+    // VISITORS
+    void visit(AstNode* nodep) override {
+        // Push a new stack entry at the start of a list, but only if the list is not a
+        // single element (this saves a lot of allocations in expressions)
+        bool singletonListStart = false;
+        if (nodep->backp()->nextp() != nodep) {  // If at head of list ...
+            singletonListStart = nodep->nextp() == nullptr;
+            if (!singletonListStart) m_stack.emplace_back(m_hasher);
+        }
+
+        // Analyse node
+        if (AstNodeStmt* const stmtp = VN_CAST(nodep, NodeStmt)) {
+            analyseStmt(stmtp, /*tryCondMatch:*/ !singletonListStart);
+        } else if (AstVarRef* const vrefp = VN_CAST(nodep, VarRef)) {
+            analyseVarRef(vrefp);
         } else {
-            return Mergeable::YES;
+            analyseNode(nodep);
         }
+
+        // Pop the stack at the end of a list
+        if (!singletonListStart && !nodep->nextp()) m_stack.pop_back();
+    }
+
+    // CONSTRUCTOR
+    CodeMotionAnalysisVisitor(AstNode* nodep,
+                              AstUser3Allocator<AstNodeStmt, StmtProperties>& stmtProperties)
+        : m_stmtProperties(stmtProperties) {
+        iterateAndNextConstNull(nodep);
+    }
+
+public:
+    static void analyse(AstNode* nodep,
+                        AstUser3Allocator<AstNodeStmt, StmtProperties>& stmtProperties) {
+        CodeMotionAnalysisVisitor{nodep, stmtProperties};
+    }
+};
+
+class CodeMotionOptimizeVisitor final : public AstNVisitor {
+    // NODE STATE
+    // AstNodeStmt::user3   -> StmtProperties (accessed via m_stmtProperties, maneged
+    //                         externally, see MergeCondVisitor::process)
+    // AstNodeStmt::user4   -> Flag: Already processed this node
+
+    AstUser4InUse m_user4InUse;
+
+    const AstUser3Allocator<AstNodeStmt, StmtProperties>& m_stmtProperties;
+
+    // MEMBERS
+
+    // Predicate to check if two sets are disjoint
+    static bool areDisjoint(const std::set<const AstVar*>& a, const std::set<const AstVar*>& b) {
+        if (a.empty() || b.empty()) return true;
+        const auto endA = a.end();
+        const auto endB = b.end();
+        auto itA = a.begin();
+        auto itB = b.begin();
+        for (;;) {
+            if (*itA == *itB) return false;
+            if (std::less<const AstVar*>{}(*itA, *itB)) {
```

Comment:
Will this lead to sort instability?  If not please add comment why.

Also generally, please see if anything else makes the algorithm now sort astable, thanks.

Reply:
Explained in comment